dm/syncer/binlogstream/stream_modifier.go

```diff
@@ -0,0 +1,252 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binlogstream
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/google/uuid"
+	"github.com/pingcap/tiflow/dm/dm/pb"
+	"github.com/pingcap/tiflow/dm/pkg/binlog"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"go.uber.org/zap"
+)
+
+// operator contains an operation for specified binlog pos
+// used by `handle-error`.
+type operator struct {
+	uuid      string // add a UUID, make it more friendly to be traced in log
+	op        pb.ErrorOp
+	pos       mysql.Position
+	events    []*replication.BinlogEvent // ddls -> events
+	originReq *pb.HandleWorkerErrorRequest
+}
+
+// newOperator creates a new operator with a random UUID.
+func newOperator(
+	op pb.ErrorOp,
+	pos mysql.Position,
+	events []*replication.BinlogEvent,
+	originReq *pb.HandleWorkerErrorRequest,
+) *operator {
+	reqClone := *originReq
+	return &operator{
+		uuid:      uuid.New().String(),
+		op:        op,
+		pos:       pos,
+		events:    events,
+		originReq: &reqClone,
+	}
+}
+
+func (o *operator) String() string {
+	events := make([]string, 0)
+	for _, e := range o.events {
+		buf := new(bytes.Buffer)
+		e.Dump(buf)
+		events = append(events, buf.String())
+	}
+	return fmt.Sprintf("uuid: %s, op: %s, events: %s, originReq: %v", o.uuid, o.op, strings.Join(events, "\n"), o.originReq)
+}
+
+// streamModifier is not thread-safe.
+type streamModifier struct {
+	ops    []*operator // sorted on operator.Position
+	nextOp int         // next operator whose location is waiting to be matched
```

Comment:
why we keep ops before `nextOp`?

Reply:
the op may need to be consumed again by sharding DDL resync. For example

position 10: inject [tbl1 DDL1, tbl1 DDL2]
position 11: inject ...
position 20: tbl2 DDL1
position 30: tbl2 DDL2

the consuming order is 10, 11, 20 (now finish sharding DDL1, resync), 10, 11, 20, 30. Assuming table checkpoint will filter DDLs of finished sharding groups.

global checkpoint is used to `RemoveOutdated`